### PR TITLE
Allows the server to serve static files

### PR DESCRIPTION
**Pull request by NigelGreenway**
Please see issue #2  for further context.

### CODE REVIEW

**Comment by mbarquin** [comment]
Hello, 

I'm working on your code, I think I can make a new ReactContainer object to be injected on server constructor, or, if it is  null, the server constructor makes a new instance, the "real" React object instances (loop, socket, http) are created in the constructor of the container, the operations to configure the react server will be be also on this new container, I will try to alias the react server operations inside the object like listen or run... as soon as I can I submit you a PR. 

After that I'm thinking about a unique run($app) method, to configure if Assets will be served we can add a new method $server->withAssets($pathToAssets); . And in the old public run($app) method switch between runAsApiServer() or runAsWebServer() as private methods

thoughts?

**Comment by NigelGreenway** [comment]
@mbarquin I think an extra object for the ReactContainer would be overkill?

In regards to the `run` method, I did do something like that originally [here](https://github.com/mbarquin/reactphp-slim/pull/3/commits/65b2ebe490de00f42bec09b4855d246518e8a39c#diff-c611f5024b5b13f69f779505d1c6735eR190) but it looked messy and annoying to read and follow. My preference would be to have a separate `public` method for each but maybe named better? 😃 

I'm open to suggestion though and would go the path you choose, as after all this is your creation 😉

File: readme.md [hunk 1]
```diff
@@ -57,6 +57,19 @@ Sets the IP to be listened to
...
 Sets the port the server will be listening to, by default it will be set to 1337.
 
+An optional method is available to serve web pages and the assets directory if required:
+
+**withWebRoot($string)**
+Sets the web root in order to serve static content. For example, if you have a `./src/app/app.php` file and you assets are located in `./src/app/assets` then you would call the following:
+
+```php
+$server
+    ->withHost('192.168.67.1')
```
Comment (NigelGreenway) [inline readme.md:67]: Needs removing/updating upon final implementation being decided

File: src/Server.php [hunk 2]
```diff
@@ -97,68 +102,137 @@ public function withHost($ip)
...
+     * Run a full web server and serve web assets as well as HTML
+     *
+     * @param App    $app
+     * @param string $webRootDirectory
+     *
+     * @return void
+     */
+    public function runAsWebServer(App $app, $webRootDirectory)
+    {
+        if (
```
Comment (NigelGreenway) [inline src/Server.php:206]: Would be better to extract into a `private` method

File: src/Server.php [hunk 3]
```diff
@@ -97,68 +102,137 @@ public function withHost($ip)
...
+            && is_dir($webRootDirectory) === false
+        ) {
+            throw new \InvalidArgumentException($webRootDirectory);
+        }
+
+        list(
+            $loop,
+            $socketServer,
+            $httpServer
+            ) = $this->initialiseReactPHP();
```
Comment (NigelGreenway) [inline src/Server.php:217]: Bad indentation

**Comment by mbarquin** [comment]
Hi, 

Another thing... 

`$server = $this;
        return function (
               \React\Http\Request $request,
               \React\Http\Response $response) use ($app, $server) {`

Is replaced by

` $this->react->getHttpServer()->on('request', function (Request $request, Response $response) use ($app) {
            $this->bootApp($app, $request, $response);
        });`

$this is used inside the anonymous callback, I'm not sure if $this (server object) will always have the right scope along future changes, I think it can be more explicit passing the server object ($this) along the **use** clause, but using $this on this clause is forbidden.... then $server = $this

What do you think?

**Comment by mbarquin** [comment]
@NigelGreenway I think switching some react native code to its own object will grant a better code quality. I think it woul be better to call it **ReactAdapter instead of ReactContainer**

ReactAdapter can implement a ServerInterface, our server class will be waiting for this interface, and in future, we will be able to write adapters for other server libraries and inject them to our own server class on constructor.

It will allow us to mock these adapters for testing.

If nor reactAdapter nor other adapter is injected on server costructor (a reactAdapter will be instanciated on this constructor, as default), 

We keep the server class with as few responsabilities as we can. I think that this structure will not make a significant overhead as PHP should be optimized for an OOP Context.

`namespace mbarquin\reactSlim\Container;

use \React\EventLoop\Factory;
use \React\Socket\Server as SocketServer;
use \React\Http\Server;

class ReactAdapter implements ServerInterface
{
    private $loop;
    private $socketServer;
    private $httpServer;

    public function __construct()
    {
        $this->loop         = Factory::create();
        $this->socketServer = new SocketServer($this->loop);

        $this->httpServer   = new Server(
                $this->socketServer, $this->loop
            );
    }

    public function getLoop()
    {
        return $this->loop;
    }

    public function getSocketServer()
    {
        return $this->socketServer;
    }

    public function getHttpServer()
    {
        return $this->httpServer;
    }

    public function listen($port, $host) {
        $this->socketServer->listen($port, $host);

        return $this;
    }

    public function run()
    {
         $this->loop->run();
    }

}`

In server class
`class server 
{

    public function __construct(ReactAdapter $erverAdpt = null)
    {
        if($erverAdpt === null) {
            $this->serverAdapter = new ReactAdapter();
        } else {
            $this->serverAdapter = $erverAdpt;
        }
    }

// ......`

thoughts?

**Comment by NigelGreenway** [comment]
Yeah, that makes sense. How best to add this? Maybe get this PR in and then another PR for this new object?

**Comment by mbarquin** [comment]
Perhaps this reactAdapter it's not exactly in the perimeter of this pull request, I've suggested it to make the PR a +1 to the total quality of the project, the code to create the react object had moved, it's a good moment for it. 

I have it coded, I must to test it and I can make a PR to your branch, when the request will be accepted it will be pushed to this PR (I think).

I'm slowly this days, I have a lot of tasks opened. I would like to merge your code as soon as possible. But now I have three doubts: 

* The first is about the use of $this inside the callback "on request" . I don't know if the system will keep a reference to the object which creates the callback or it can be misunderstanded depending on context. The callback itself is called out of the object which creates it (server), the callable written I think it isn't part of server Object.

* The second is about the closures I were using, I'm not sure if they were redundant... I'll try to check it tonight.

* The third is about the method run vs "apiRun" + "webRun", I'll make a mockup to show you what I think and I will query some colleagues to see different opinions.

If you are in a hurry tell me, and I will try to make the flow a bit faster.... Thanks for all

**Comment by NigelGreenway** [comment]
What if we decline this PR and look at getting the other parts cleaned up like adding the PHP7 features like scalar type hints, return type hints and strict mode? I can revert the commit that removes `$server = $this;`, although I've tested the scope (in a basic test) and it's returning the parent class (`Server`).

If we get #4 merged with all the PHP 7 features added then look at getting an issue in as an RFC where we can discuss the best implementation maybe?

**Comment by mbarquin** [comment]
Hi, 

Ok, I can accept the php7 modification and after we can talk a bit about the assets, I think we can make a first quick implementation and after that depending on what you want we can make a good implementation, I was thinking also on a good way to configure which files should be considered assets...

On the other hand, after digging a bit I found this entry in stackoverflow, you were right, there is no reason to assing $this to $server.

[http://stackoverflow.com/questions/8391099/using-this-in-anonymous-function](http://stackoverflow.com/questions/8391099/using-this-in-anonymous-function)

**Comment by NigelGreenway** [comment]
Yeah, that makes sense.

I was playing around with that last night and found php5.4 is the first time `$this` is supported within a closure.

**Comment by NigelGreenway** [comment]
Are you okay with this being closed?

**Comment by NigelGreenway** [comment]
@mbarquin 

> Are you okay with this being closed?

**Pull request closed without merging**